scripts/memfault/mds_ble_gateway.py

```diff
@@ -0,0 +1,399 @@
+            if status[0] == BLEGattStatusCode.success:
+                data.extend(status[1])
+
+        project_key = "".join(chr(c) for c in data)
+
+        return project_key
+
+    def data_export_notification_enable(self):
+        self.adapter.enable_notification(conn_handle=self.conn_handle, uuid=BLEMemfault.DATA_EXPORT)
+
+    def data_export_write(self, data):
+        self.adapter.write_req(self.conn_handle, BLEMemfault.DATA_EXPORT, data)
+
+    def close(self):
+        self.adapter.close()
+
+    def on_gap_evt_adv_report(self, ble_driver, conn_handle, peer_addr, rssi, adv_type, adv_data):
+        uuid = 0
+
+        if BLEAdvData.Types.service_128bit_uuid_complete in adv_data.records:
+            uuid = adv_data.records[BLEAdvData.Types.service_128bit_uuid_complete]
+            uuid.reverse()
+
+            logger.info('BLE: Received advertising packet with the MDS UUID')
+
+        address = "".join(f'{b:02X}:' for b in peer_addr.addr)
+        address = address[:-1]
+
+        logger.debug(f'BLE: Received advertisement report, address: {address}')
+
+        if uuid == MDS_UUID:
+            conn_params = BLEGapConnParams(min_conn_interval_ms = 15,
+                                           max_conn_interval_ms = 30,
+                                           conn_sup_timeout_ms  = 4000,
+                                           slave_latency        = 0)
+
+            logger.info(f'BLE: Connecting to {address}')
+            self.adapter.connect(address=peer_addr, conn_params=conn_params,
+                                 tag=BLEMemfault.CFG_TAG)
+
+    def on_gap_evt_connected(self, ble_driver, conn_handle, peer_addr, role, conn_params):
+        self.evt_sync.notify('connected', data=conn_handle)
+
+    def on_gap_evt_disconnected(self, ble_driver, conn_handle, reason):
+        self.evt_sync.notify('disconnected', data=conn_handle)
+        logger.info(f'BLE: Disconnected (reason: {reason})')
+
+    def on_gap_evt_sec_request(self, ble_driver, conn_handle, bond, mitm, lesc, keypress):
+        logger.info(f'BLE: Security request, bond: {bond}')
+
+    def on_gap_evt_conn_sec_update(self, ble_driver, conn_handle, conn_sec):
+        logger.info(f'BLE: Security changed level: {conn_sec.sec_mode.lv}')
+
+    def on_gap_evt_auth_status(self, ble_driver, conn_handle, **kwargs):
+        status = kwargs.get('auth_status')
+        if status == BLEGapSecStatus.success:
+            status = 'Success'
+        else:
+            status = 'Failed'
+
+        logger.info(f'BLE: Authentication status conn_handle: {conn_handle} status: {status}')
+
+    def on_notification(self, ble_adapter, conn_handle, uuid, data):
+        if conn_handle != self.conn_handle:
+            return
+
+        if uuid != BLEMemfault.DATA_EXPORT:
+            return
+
+        logger.debug(f'Received data export notification, data length: {len(data)}')
+
+        self.data_export_received(data)
+
+
+class Memfault(BLEMemfault):
```

Comment:
Consider adding comments for classes and functions.

Reply:
Ok, I added a docstring for main classes and functions